File: app/src/main/java/me/tagavari/airmessage/util/ConversationValueUpdateResult.java
package me.tagavari.airmessage.util;

import java.util.ArrayList;
import java.util.List;

import me.tagavari.airmessage.messaging.ConversationInfo;
import me.tagavari.airmessage.redux.ReduxEmitterNetwork;
import me.tagavari.airmessage.redux.ReduxEventMessaging;

public class ConversationValueUpdateResult {
	private final boolean unarchived;
	private final int unreadIncrement;
	
	public ConversationValueUpdateResult(boolean unarchived, int unreadIncrement) {
		this.unarchived = unarchived;
		this.unreadIncrement = unreadIncrement;
	}
	
	public boolean isUnarchived() {
		return unarchived;
	}
	
	public int getUnreadIncrement() {
		return unreadIncrement;
	}
	
	public List<ReduxEventMessaging> getEvents(ConversationInfo conversationInfo) {
		List<ReduxEventMessaging> events = new ArrayList<>(2);
		if(unarchived) events.add(new ReduxEventMessaging.ConversationArchive(conversationInfo, false));
		if(unreadIncrement > 0) events.add(new ReduxEventMessaging.ConversationUnread(conversationInfo, conversationInfo.getUnreadMessageCount() + unreadIncrement));
		return events;
	}
	
	public void emitUpdate(ConversationInfo conversationInfo) {
		//Emitting updates
		if(unarchived) ReduxEmitterNetwork.getMessageUpdateSubject().onNext(new ReduxEventMessaging.ConversationArchive(conversationInfo, false));
		if(unreadIncrement > 0) ReduxEmitterNetwork.getMessageUpdateSubject().onNext(new ReduxEventMessaging.ConversationUnread(conversationInfo, conversationInfo.getUnreadMessageCount() + unreadIncrement));
	}
}

File: app/src/main/java/me/tagavari/airmessage/util/ConversationValueUpdateResult.kt
package me.tagavari.airmessage.util

import me.tagavari.airmessage.messaging.ConversationInfo
import me.tagavari.airmessage.redux.ReduxEmitterNetwork
import me.tagavari.airmessage.redux.ReduxEventMessaging
import me.tagavari.airmessage.redux.ReduxEventMessaging.ConversationArchive
import me.tagavari.airmessage.redux.ReduxEventMessaging.ConversationUnread

/**
 * Represents an update to a conversation due to an incoming message update
 * The conversation may be unarchived due to the incoming message, and will have its unread count incremented
 */
class ConversationValueUpdateResult(private val isUnarchived: Boolean, private val unreadIncrement: Int) {
	fun getEvents(conversationInfo: ConversationInfo): List<ReduxEventMessaging> {
		val events = mutableListOf<ReduxEventMessaging>()
		if(isUnarchived) events.add(ConversationArchive(conversationInfo, false))
		if(unreadIncrement > 0) events.add(ConversationUnread(conversationInfo, conversationInfo.unreadMessageCount + unreadIncrement))
		return events
	}
	
	fun emitUpdate(conversationInfo: ConversationInfo) {
		//Emitting updates
		if(isUnarchived) ReduxEmitterNetwork.messageUpdateSubject.onNext(ConversationArchive(conversationInfo, false))
		if(unreadIncrement > 0) ReduxEmitterNetwork.messageUpdateSubject.onNext(ConversationUnread(conversationInfo, conversationInfo.unreadMessageCount + unreadIncrement))
	}
}